### ISSUE TITLE
Integration in TypeScript project broke again

### ISSUE THREAD

**Issue by KingHenne** [rollup/rollup#1902]
When upgrading rollup from `0.54.1` to `0.55.0`in [a TypeScript project](https://github.com/KingHenne/declare-react-module) I get new TypeScript compiler errors (see also #1862):

```
node_modules/rollup/dist/typings/Graph.d.ts(1,22): error TS6053: File '/Users/hendrik/code/open-source/declare-react-module/node_modules/rollup/src/Graph.d.ts' not found.

1 /// <reference path="../../src/Graph.d.ts" />
                       ~~~~~~~~~~~~~~~~~~~~


node_modules/rollup/dist/typings/watch/index.d.ts(2,8): error TS1192: Module '"events"' has no default export.

2 import EventEmitter from 'events';
         ~~~~~~~~~~~~
```

At this point maybe a test could be helpful that compiles a tiny TS project that consumes the typings in `dist/typings` to catch those errors before publishing a new version. I could contribute that...

**Comment by jthoms1** [comment]
I am experiencing this as well.  It appears that 0.55.0 has shipped some incorrect type info.  I am researching the fix. I am assuming that some of these errors are because of dev dependencies that are rolled into the final and the types are still expecting them to be installed as dependencies.


```
node_modules/rollup/dist/typings/Graph.d.ts(1,22): error TS6053: File '/Users/joshthomas/Workspace/stencil/node_modules/rollup/src/Graph.d.ts' not found.

1 /// <reference path="../../src/Graph.d.ts" />
                       ~~~~~~~~~~~~~~~~~~~~


node_modules/rollup/dist/typings/Graph.d.ts(2,23): error TS2688: Cannot find type definition file for 'acorn'.

2 /// <reference types="acorn" />
                        ~~~~~


node_modules/rollup/dist/typings/Graph.d.ts(3,24): error TS7016: Could not find a declaration file for module 'acorn'. '/Users/joshthomas/Workspace/stencil/node_modules/acorn/dist/acorn.js' implicitly has an 'any' type.
  Try `npm install @types/acorn` if it exists or add a new declaration (.d.ts) file containing `declare module 'acorn';`

3 import * as acorn from 'acorn';
                         ~~~~~~~


node_modules/rollup/dist/typings/Module.d.ts(5,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/joshthomas/Workspace/stencil/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm install @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`

5 import { RawSourceMap } from 'source-map';
                               ~~~~~~~~~~~~


node_modules/rollup/dist/typings/rollup/index.d.ts(3,30): error TS7016: Could not find a declaration file for module 'source-map'. '/Users/joshthomas/Workspace/stencil/node_modules/source-map/source-map.js' implicitly has an 'any' type.
  Try `npm install @types/source-map` if it exists or add a new declaration (.d.ts) file containing `declare module 'source-map';`

3 import { RawSourceMap } from 'source-map';
                               ~~~~~~~~~~~~


node_modules/rollup/dist/typings/watch/index.d.ts(4,30): error TS7016: Could not find a declaration file for module 'chokidar'. '/Users/joshthomas/Workspace/stencil/node_modules/chokidar/index.js' implicitly has an 'any' type.
  Try `npm install @types/chokidar` if it exists or add a new declaration (.d.ts) file containing `declare module 'chokidar';`

4 import { WatchOptions } from 'chokidar';
```

**Comment by KingHenne** [comment]
The first issue is solved in #1903. For the last three issues see https://rollupjs.org/guide/en#typescript-declarations, you have to install `@types/chokidar`, `source-map` and `magic-string`. The issue with `/// <reference types="acorn" />` is interesting since I don't see that line being emitted. Anyways, it seems that `acorn` needs to be added as dependency as well and the [section linked to above](https://rollupjs.org/guide/en#typescript-declarations) should be updated.

**Comment by jthoms1** [comment]
I just installed the types for acorn as a dev dependency as well and the error resolved. Thanks again. `npm install --save-dev @types/acorn`

**Comment by guybedford** [comment]
I'll stress my opinion here again - we shouldn't automatically generate the types, we should just manually maintain the simple interface types... Rollup doesn't have a huge surface area to cover here and the automated type build seems overly complex.

**Comment by jthoms1** [comment]
I agree with this.  Having to npm install the extra types seems really odd.  I have worked with quite a few typed libs and this is a first for me.  I think the process could be much simpler.

**Comment by KingHenne** [comment]
I agree as well, having the separate `tsconfig-types.json` seems to be a really odd approach. I wonder wether the emitted d.ts files would be better consumable if it were a standard TypeScript setup. I will try to investigate this a bit...

**Comment by KingHenne** [comment]
> I agree as well, having the separate `tsconfig-types.json` seems to be a really odd approach. I wonder wether the emitted d.ts files would be better consumable if it were a standard TypeScript setup. I will try to investigate this a bit...

I'm not optimistic we can improve anything here. 🙁 ((@guybedford you probably know this already)

**Comment by KingHenne** [comment]
@guybedford After looking a bit into it I'm not sure you can even solve the dependency issue by manually crafting the typings for rollup. You would still end up with references to acorn types for example. I also remembered, that I did exactly that a few months ago: https://github.com/KingHenne/declare-react-module/blob/v1.5.0/typings/rollup.d.ts

One solution could be to contribute manually created typings (or maybe even the generated ones) to https://github.com/DefinitelyTyped/DefinitelyTyped and declare the dependencies there. So then `rollup` would stay without dependencies and instead`@types/rollup` would have the dependencies and thus consuming TypeScript projects wouldn't need to manually install them.

**Comment by alan-agius4** [comment]
I just looked at the 2 problem you are having and both errors are being encountered because of how the TypeScript source is written;

https://github.com/rollup/rollup/blob/master/src/Graph.ts#L1

and 

https://github.com/rollup/rollup/blob/master/src/watch/index.ts#L2

are not correct and how there is used and declared will certainly cause problems for consumers.

**Comment by KingHenne** [comment]
@alan-agius4 I know, that's why I fixed both issues in #1903, please take a look.

**Comment by dominique-mueller** [comment]
I'm curious, `preserveSymlinks` within the `InputOptions` is now required by the TypeScript typings - but wasn't before. Is this intentional?

**Comment by KingHenne** [comment]
@dominique-mueller This has been fixed in #1939.
